libxrdp/xrdp_sec.c

```diff
@@ -2302,159 +2302,235 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
-xrdp_sec_process_mcs_data_monitors(struct xrdp_sec *self, struct stream *s)
+int
+process_monitor_stream(struct stream *s, struct display_size_description *description, int full_parameters)
 {
-    int index;
-    int monitorCount;
-    int flags;
-    int x1;
-    int y1;
-    int x2;
-    int y2;
-    int got_primary;
-    struct xrdp_client_info *client_info = (struct xrdp_client_info *)NULL;
-
-    client_info = &(self->rdp_layer->client_info);
-
-    /* this is an option set in xrdp.ini */
-    if (client_info->multimon != 1) /* are multi-monitors allowed ? */
-    {
-        LOG(LOG_LEVEL_INFO, "Multi-monitor is disabled by server config");
-        return 0;
-    }
-    if (!s_check_rem_and_log(s, 8, "Parsing [MS-RDPBCGR] TS_UD_CS_MONITOR"))
+    LOG_DEVEL(LOG_LEVEL_TRACE, "process_monitor_stream:");
+    if (description == NULL)
     {
+        LOG_DEVEL(LOG_LEVEL_DEBUG, "\tprocess_monitor_stream: description was null. Valid pointer to allocated description expected.");
         return 1;
     }
-    in_uint32_le(s, flags); /* flags */
-    in_uint32_le(s, monitorCount);
-    LOG_DEVEL(LOG_LEVEL_TRACE, "Received [MS-RDPBCGR] TS_UD_CS_MONITOR "
-              "flags 0x%8.8x, monitorCount %d", flags, monitorCount);
 
-    //verify flags - must be 0x0
-    if (flags != 0)
+    int num_monitor;
+    struct monitor_info *monitor_layout;
+    struct xrdp_rect rect = {8192, 8192, -8192, -8192};
+    int got_primary = 0;
+    int monitor_struct_stream_check_bytes;
+    const char *monitor_struct_stream_check_message;
+
+    in_uint32_le(s, num_monitor);
+    LOG(LOG_LEVEL_DEBUG, "  num_monitor %d", num_monitor);
+
+    if (num_monitor >= CLIENT_MONITOR_DATA_MAXIMUM_MONITORS)
     {
         LOG(LOG_LEVEL_ERROR,
-            "[MS-RDPBCGR] Protocol error: TS_UD_CS_MONITOR flags MUST be zero, "
-            "received: 0x%8.8x", flags);
+            "\tprocess_monitor_stream: [MS-RDPBCGR] Protocol error: TS_UD_CS_MONITOR monitorCount "
+            "MUST be less than 16, received: %d", num_monitor);
         return 1;
     }
-    //verify monitorCount - max 16
-    if (monitorCount > 16)
+
+    /*
+     * Unfortunately the structure length values aren't directly defined in the Microsoft specifications.
+     * They are derived from the lengths of the specific structures referenced below.
+     */
+    if (full_parameters == 0)
     {
-        LOG(LOG_LEVEL_ERROR,
-            "[MS-RDPBCGR] Protocol error: TS_UD_CS_MONITOR monitorCount "
-            "MUST be less than 16, received: %d", monitorCount);
-        return 1;
+        monitor_struct_stream_check_bytes = 20;
+        monitor_struct_stream_check_message = "\tprocess_monitor_stream: Parsing monitor definitions from 2.2.1.3.6.1 Monitor Definition (TS_MONITOR_DEF).";
+    }
+    else
+    {
+        monitor_struct_stream_check_bytes = 40;
+        monitor_struct_stream_check_message = "\tprocess_monitor_stream: Parsing monitor definitions from 2.2.2.2.1 DISPLAYCONTROL_MONITOR_LAYOUT.";
```

Comment:
Since the RDP spec has many extension specs, can you please include the spec name abbreviation in the message. Eg. [MS-RDPBCGR]